packages/components/src/internal/query/api.ts

```diff
@@ -957,17 +963,60 @@ export function processRequest(response: any, request: any, reject: (reason?: an
  * provided by `@labkey/components`.
  * @private
  */
-export function getContainerFilter(containerPath?: string): Query.ContainerFilter {
-    // Check experimental flag to see if cross-folder data support is enabled.
-    if (!isProductProjectsEnabled()) {
+export function getContainerFilter(containerPath?: string, moduleContext?: ModuleContext): Query.ContainerFilter {
+    // Check to see if product projects support is enabled.
+    if (!isProductProjectsEnabled(moduleContext)) {
+        return undefined;
+    }
+
+    // When all folder filtering is enabled resolve data across all folders.
+    if (isAllProductFoldersFilteringEnabled(moduleContext)) {
+        return Query.ContainerFilter.allInProjectPlusShared;
+    }
+
+    // When requesting data from a top-level folder context the ContainerFilter filters
+    // "down" the folder hierarchy for data.
+    if (isProjectContainer(containerPath)) {
+        return Query.ContainerFilter.currentAndSubfoldersPlusShared;
+    }
+
+    // When requesting data from a sub-folder context the ContainerFilter filters
+    // "up" the folder hierarchy for data.
+    return Query.ContainerFilter.currentPlusProjectAndShared;
+}
+
+/**
+ * Provides the configured ContainerFilter to utilize when requesting data that is being read
+ * within a folder context.
+ * @private
+ */
+export function getContainerFilterForFolder(
+    containerPath?: string,
+    moduleContext?: ModuleContext
+): Query.ContainerFilter {
+    // Check to see if product projects support is enabled.
+    if (!isProductProjectsEnabled(moduleContext)) {
         return undefined;
     }
 
-    const isProject = isProjectContainer(containerPath);
+    if (isProductProjectsDataListingScopedToProject(moduleContext)) {
+        // When requesting data from a top-level folder context the ContainerFilter filters
+        // "down" the folder hierarchy for data.
+        if (isProjectContainer(containerPath)) {
+            if (isAllProductFoldersFilteringEnabled(moduleContext)) {
+                return Query.ContainerFilter.allInProjectPlusShared;
+            }
+            return Query.ContainerFilter.currentPlusProjectAndShared;
+        }
+
+        // When listing data in a folder scope returned data to the current
+        // folder when the experimental feature is enabled.
+        return Query.ContainerFilter.current;
```

Comment:
all of these options will definitely make for interesting combinations of test cases to make sure we keep working as expected. do you know if all clients will be headed in this direction for the new project behavior or if we'll have to support both/all options long term?

Reply:
Indeed. There is a long-term vision for projects, laid out by Hannah, that will introduce further significant changes to projects. We're introducing these features under experimental flags with the intent to either switch to their paradigm completely OR expose these as configuration options. Hopefully, we can find a happy medium where we allow configuration while not having it be as mentally taxing for us to work on / deliver stability.